### PR TITLE
azfile: File upload methods

### PR DESCRIPTION
**Pull request by souravgupta-msft**
- [x] The purpose of this PR is explained in this or a referenced issue.
- [x] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [x] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/storage/azfile/file/chunkwriting.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	// It returns the total number of buffers or an error.
+	// No error is returned if the number of buffers has reached max.
+	// This is called only from the reading goroutine.
+	Grow() (int, error)
+
+	// Free cleans up all buffers.
+	Free()
+}
+
+// copyFromReader copies a source io.Reader to blob storage using concurrent uploads.
```
Comment (gapra-msft) [inline sdk/storage/azfile/file/chunkwriting.go:43]:
```suggestion
// copyFromReader copies a source io.Reader to file storage using concurrent uploads.
```

File: sdk/storage/azfile/file/client.go [hunk 2]
```diff
@@ -266,3 +271,114 @@ func (f *Client) GetSASURL(permissions sas.FilePermissions, expiry time.Time, o
...
+
+func (f *Client) UploadBuffer(ctx context.Context, buffer []byte, options *UploadBufferOptions) error {
+	uploadOptions := uploadFromReaderOptions{}
+	if options != nil {
+		uploadOptions = *options
+	}
+	return f.uploadFromReader(ctx, bytes.NewReader(buffer), int64(len(buffer)), &uploadOptions)
+}
+
+// UploadFile uploads a file in blocks to a block blob.
```
Comment (tasherif-msft) [inline sdk/storage/azfile/file/client.go:335]:
```suggestion
// UploadFile uploads a file
```

File: sdk/storage/azfile/file/chunkwriting.go (hunk 1 above)
Comment (souravgupta-msft) [reply sdk/storage/azfile/file/chunkwriting.go]: Done

File: sdk/storage/azfile/file/client.go (hunk 2 above)
Comment (souravgupta-msft) [reply sdk/storage/azfile/file/client.go]: Done

File: sdk/storage/azfile/internal/shared/shared.go [hunk 3]
```diff
@@ -164,3 +170,53 @@ func GenerateLeaseID(leaseID *string) (*string, error) {
...
+type nopCloser struct {
+	io.ReadSeeker
+}
+
+func (n nopCloser) Close() error {
+	return nil
+}
+
+// NopCloser returns a ReadSeekCloser with a no-op close method wrapping the provided io.ReadSeeker.
+func NopCloser(rs io.ReadSeeker) io.ReadSeekCloser {
```
Comment (gapra-msft) [inline sdk/storage/azfile/internal/shared/shared.go:220]: Doesnt azcore provide a NopCloser in the streaming package?

**Review by vibhansa-msft — APPROVED** [review]

**Review by tasherif-msft — APPROVED** [review]

File: sdk/storage/azfile/internal/shared/shared.go (hunk 3 above)
Comment (souravgupta-msft) [reply sdk/storage/azfile/internal/shared/shared.go]: Thanks for pointing this. I have removed this part from azfile. This should be removed from azblob as well.

**Pull request merged**